src/collections.h

```diff
@@ -119,6 +200,39 @@ Collection<T>::fold(std::function<T(T, T)> func) {
 
 
 template<typename T>
+template<typename ...U>
+Collection<std::tuple<U...>>
+Collection<T>::zip(Collection<U>... other_list) {
+    // TODO: list size checking
+    using return_type = std::tuple<U...>;
+
+    std::vector<return_type> list;
+    for (int i = 0, e = std::min({other_list.size()...}); i < e; i++) {
+        list.emplace_back(std::move(other_list.front())...);
+        [](...){} ((other_list.pop_front(), 0)...); 
+    }
+    return Collection<return_type>(list);
+};
+
+
+template<typename T>
+template<typename Function, typename ...U>
+Collection<typename std::result_of<Function(U...)>::type>
+Collection<T>::zipWith(Function func, Collection<U>... other_list) {
+    // TODO: list size checking
+    // TODO: check that func takes as many arguments as there are lists
+    using return_type = typename std::result_of<Function(U...)>::type;
+
+    std::vector<return_type> list;
```

Comment:
we should find the min_size and store it in a variable so that we can instantiate the list to the proper size.